contxt/services/iot.py

```diff
@@ -114,6 +125,79 @@ def get_time_series_for_field(
             per_page=per_page,
         )
 
+    def get_time_series_for_fields(
+        self,
+        fields: List[Field],
+        start_time: datetime,
+        window: Window = Window.RAW,
+        end_time: Optional[datetime] = None,
+    ) -> List[FieldTimeSeries]:
+        """Get complete (non-paginated) time series data for each field in `fields`"""
+        # Build requests queue
+        params = {
+            "timeStart": int(start_time.timestamp()),
+            "timeEnd": int(end_time.timestamp()) if end_time else None,
+            "window": window.value,
+            "limit": 5000,
+        }
+        queue: List[Tuple[str, BatchRequest]] = [
+            (
+                f.field_human_name,
+                BatchRequest.from_request(
+                    Request(
+                        method="GET",
+                        url=self._url(
+                            f"outputs/{f.output_id}/fields/{f.field_human_name}/data"
+                        ),
+                        params=params,
+                    )
+                ),
+            )
+            for f in fields
+        ]
+
+        # Make all requests in queue
+        MAX_BATCH_REQUESTS = 200
+        records: Dict[str, Dict[datetime, str]] = defaultdict(dict)
+        while queue:
+            # Make next batch of requests
+            requests = dict(queue[:MAX_BATCH_REQUESTS])
+            del queue[:MAX_BATCH_REQUESTS]
+            logger.info(f"Making {len(requests)} batched requests to IOT API")
+            responses = self._batch_request(requests)
+            any_success = False
+
+            # Process responses
+            for name, resp in responses.items():
+                if resp.ok:
+                    any_success = True
+                    series = {
+                        Parsers.datetime(r["event_time"]): Parsers.unknown(r["value"])
+                        for r in resp.body["records"]
+                    }
+                    records[name].update(series)
+                    # Add request for next page
+                    next_page_url = resp.body["meta"]["next_page_url"]
+                    if next_page_url:
+                        queue.append(
+                            (name, BatchRequest(method="GET", uri=next_page_url))
+                        )
+                else:
+                    # Retry request
+                    logger.warning(
+                        f"Got bad response from IOT API ({resp.body}). Retrying..."
+                    )
+                    queue.append((name, requests[name]))
+
+            if not any_success:
+                raise IOError(f"All {len(requests)} batched requests to IOT API failed")
+
+        fields_by_name = {f.field_human_name: f for f in fields}
+        return [
+            FieldTimeSeries(field=fields_by_name[name], time_series=series)
+            for name, series in records.items()
+        ]
+
```

Comment:
Just a suggestion, but does using something like async/await from asyncio make sense to use here instead of the queueing?